### DISCUSSION TITLE
Adding a new key/value in session?

### DISCUSSION THREAD

**Discussion by soheil-bl** [nextauthjs/next-auth#5497] [Help]
### Question 💬

Hi there,

TL;DR:
Is there a way to save additional information (e.g. key/value pair) in next-auth session or JWT token at some point time after the sign in flow completed?

Context:
1. I'm working on a web app with multi-tenancy. I have separately modeled Users and Orgs in DB.
2. A signed in user (e.g. keyed based on user email address) can potentially belong to multiple Orgs. 
3. If a user is a member of multiple Orgs, the user needs to tell the app which Org they would like to load in the app. This is done via an "organization picker" UI, where the user selects which organization context it would like to use the app under. Note that this UI will be shown to the user at some point after the initial signin()/session()/jwt() callbacks from next-auth is invoked
4. Once this selection is made, I would ideally like to store the selected OrgId in next-auth's session or JWT for future retrieval. This will help the server identify which Org's data it should return in the api responses.

What I've researched so far: 
1. I understand that there are some callbacks e.g. session() and jwt() that could be configured. 
https://next-auth.js.org/configuration/callbacks#jwt-callback

2. I also know that calling the getSession(), useSession(), or unstable_getServerSession() would result in the session() and jwt() callbacks being invoked. However, I do not see any way of passing along an additional key/value pair as a param that could be passed down to session() or jwt() callbacks so I can save them for future retrieval.

Questions
1. Am I missing something obvious?
1. Is my approach fundamentally wrong for wanting to store additional info in jwt/session tokens after signin flow has passed? if so, would you recommend using a plain cookie for storing such info, instead of tweaking the next-auth session/token?

Any pointers to documentation or other github threads is much appreciated!

Thanks!

### How to reproduce ☕️

N/A

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by ThangHuuVu** [comment]
> Is there a way to save additional information (e.g. key/value pair) in next-auth session or JWT token at some point time after the sign in flow completed?


This is exactly what callbacks are made for, you can use jwt and session callbacks to pass the info to the client. Did this not working for you at all?

**Reply by soheil-bl** [reply]
Thanks for the response. Well, I guess my question is:
- how can I pass a custom parameter to the jwt() or session() callbacks?

As I understand from the interface, they have a fixed set of parameters that are static:

`
jwt: (params: {
        token: JWT;
        user?: User;
        account?: A;
        profile?: P;
        isNewUser?: boolean;
    }) => Awaitable<JWT>;

session: (params: {
        session: Session;
        user: User;
        token: JWT;
    }) => Awaitable<Session>;
`

For more context, here are the walkthrough of the steps that happens during the "Sign in" and "org picker" flows:

1. User visits OAuth provider (e.g. AWS Cognito) login page.
2. OAuth provider redirects to next-auth url and passes a token.
3. Next-auth jwt() and session() callbacks get invoked. Here the server can store additional info from the logged in user token provided by OAuth provider in the jwt token. (e.g. I store username and userid in jwt token)
4. User is redirected to home page.
5. The server processing the request for home page notices that the jwt token provided by the client does not any "picked organization contexts".  So it redirects the user to the ".../org-picker" URL. The client shows the user a list of organizations that the user is a part of, so that they can pick which context they want to log in as.
6. The user picks one of the organizations. This results in the client sending a request to server with the "picked organizationId".
7. Now here, I would like to store this "picked organizationId" in jwt or session going forward. However, I don't see any way of passing this value to the jwt() or session() callbacks. As far as I could tell, there are 3 ways of invoking these callbacks: getSession(), useSession(), and unstable_getServerSession().  But the problem is, none of these accept a parameter that could be passed down to the jwt() or session() callbacks.

Does that help clarify the situation?

**Reply by ThangHuuVu** [reply]
I see, so you basically want to update the JWT/session after signing in. 🤔 
I don't think there's an elegant way to do it yet, but you can try to send your own session request to NextAuth.js that contains the `organizationId`. Try using [Advanced Initialization](https://next-auth.js.org/configuration/initialization#advanced-initialization), it allows you to handle the Auth requests before passing down to NextAuth.js.
So the flow would be: 
1,2,3,4,5,6: like above
7. After obtaining the `organizationId`, send it to `/api/auth/session` using `fetch`
8. set it in `jwt` callback

```js

export default async function auth(req: NextApiRequest, res: NextApiResponse) {
  // find organizationId
  const organizationId = findOrg() 
  return await NextAuth(req, res, {
    callbacks: {
      async jwt({ token, account, profile }) {
        // set the organizationId here
        if (account) {
          token.organizationId = organizationId
        }
        return token
      }
    }
  })
}
```

LMK if this helps 🙌

**Reply by soheil-bl** [reply]
It works thanks! Sorry for the delay.

One thing I noticed though is that this apparently only works with HTTP GET requests (not POST), and that I had to disable `useSecureCookies: false` (https://next-auth.js.org/configuration/options#usesecurecookies). 

Do you have any insight into why that is?  I thought that might be because my development server does not support HTTP, but wasn't sure.

Any pointers around that is appreciated! 🙏🏻

**Reply by ThangHuuVu** [reply]
Did it work if you're on the production server?
